### ISSUE TITLE
Solutions to CH14 Exercises

### ISSUE THREAD

**Issue by rikkuporta** [ageron/handson-ml#93]
Hi!

When will the Solutions be available?

**Comment by ageron** [comment]
Hi,

Good question, I'm struggling a bit right now (consulting + courses + several translations underway + 3 kids).  But I should be able to find some time to work on the exercise solutions within the next couple of weeks. Still missing are chapters 3, 7, 12, 14, 15, 16.  I will work first on the ones that have been requested most (14 and 16).  My apologies for the delay.

I hope you are enjoying the book!

**Comment by ageron** [comment]
I had started solutions for chapter 3, so I finished that (it's now pushed), and I just started solutions for chapter 14. I'm aiming to finish the reber grammar rnn today, I'll push the exercises as I finish them.

**Comment by rikkuporta** [comment]
Hi! Thank ypou very much - looking forward to it! :)

**Comment by ageron** [comment]
My pleasure! I just uploaded the solution to exercise 7. I'm working on exercise 8.

**Comment by StewartMenday** [comment]
First off, great book, read it cover to cover.
The solutions to the exercises seems to have stalled.  Do you have a timeline for when chapter 16 will be finished?

Cheers
Stew

**Comment by ageron** [comment]
Hi @StewartMenday ,
Yes, I'm really sorry about that, there are still 13 missing solutions in the book: 7.8, 7.9, 12.8, 12.9, 12.10, 14.8, 14.9, 15.8, 15.9, 15.10, 16.8, 16.9 and 16.10.  I'm doing my best to upload the missing solutions whenever I have some time: any particular exercise you'd like me to prioritize?

**Comment by StewartMenday** [comment]
Ageron
Thanks for getting back to me. I understand that life can get busy, so there is no pressure to get it done but when you do, I am  most interested in 16.8.

**Comment by ageron** [comment]
Hi @StewartMenday, I pushed a basic solution to exercise 16.8: the walker walks! :)
I implemented a very basic solution: simply discretizing the action space. This works, but it is extremely slow to train.  A better solution would be to use an actor-critic algorithm. It is composed of two neural nets: one "actor" network takes the state as input and outputs the action (in continuous space). The second network is the "critic": it takes both the state and the action selected by the actor as inputs, and it outputs a q-value.  Check out this nice [blog post](https://towardsdatascience.com/reinforcement-learning-w-keras-openai-actor-critic-models-f084612cfd69) for more details.
Hope this helps.
